configuring-and-sharing-volumes: Adding part3 (#212)
foundriesio/docs: source/tutorials/configuring-and-sharing-volumes/sharing-folder.rst

running containers will be able to see and consume 
the files inside it. This is one option for the host machine and the container to interact.

In the section, ":ref:`tutorial-configuring-and-sharing-volumes-using-docker`", you created and copied the configuration file to
the ``/home/shellhttpd/`` folder. Purposefully leave the changes you just did because 
you will share the same folder to see what happens.

Edit the file ``docker-compose.yml`` adding the parameter volume:

.. prompt:: bash host:~$, auto

    host:~$ gedit shellhttpd/docker-compose.yml

**shellhttpd/docker-compose.yml**:

.. prompt:: text

     version: '3.2'
     
     services:
       httpd:
         image: hub.foundries.io/<factory>/shellhttpd:latest
     #    image: shellhttpd:1.0
         restart: always
         volumes:
           - /var/rootdirs/home/fio/shellhttpd:/home/shellhttpd/
         ports:
           - 8080:${PORT-8080}
         environment:
           MSG: "${MSG-Hello world}"

Above, you add the ``volumes:`` stanza with the value  
``/var/rootdirs/home/fio/shellhttpd:/home/shellhttpd/``.
This means that the folder ``/var/rootdirs/home/fio/shellhttpd`` from your host machine 

> will be mounted over the container folder ``/home/shellhttpd``.

Reviewer: - "the container('s)" [name] "folder"
```
will be mounted over the container's ``/home/shellhttpd`` folder.
```

``Dockerfile`` will **disappear** in this new setup.

Check your changes, add, commit and push to the server:

.. prompt:: bash host:~$, auto

    host:~$ git status
    host:~$ git add shellhttpd/docker-compose.yml
    host:~$ git commit -m "Adding shared folder"
    host:~$ git push

Make sure you received your update by checking the latest **Target** on the :guilabel:`Devices` tab 
in your Factory.

Open one terminal connected to your device and check the running containers:

.. prompt:: bash device:~$, auto

    device:~$ docker ps

**Example Output**:

.. prompt:: text

     CONTAINER ID   IMAGE                               COMMAND                  CREATED          STATUS          PORTS                    NAMES
     a2d425490201   hub.foundries.io/<factory>/shellhttpd   "/usr/local/bin/http…"   20 minutes ago   Up 20 minutes   0.0.0.0:8080->8080/tcp   shellhttpd_httpd_1

Monitor the docker logs:

.. prompt:: bash device:~$, auto

    device:~$ docker logs --follow shellhttpd_httpd_1

**Example Output**:

.. prompt:: text

     PORT=8080
     MSG=Hello world

Open a second terminal connected to your device and check for the folder 

> ``/var/rootdirs/home/fio/shellhttpd``:

Reviewer: - "the" [name] "folder"
```
Open a second terminal connected to your device and check for the
``/var/rootdirs/home/fio/shellhttpd`` folder:
```

Thus, as the host machine folder is empty, the ``shellhttpd.conf`` you copied in the 
``Dockerfile`` will **disappear** in this new setup.

Check your changes, add, commit and push to the server:

.. prompt:: bash host:~$, auto

    host:~$ git status
    host:~$ git add shellhttpd/docker-compose.yml
    host:~$ git commit -m "Adding shared folder"
    host:~$ git push

Make sure you received your update by checking the latest **Target** on the :guilabel:`Devices` tab 
in your Factory.

Open one terminal connected to your device and check the running containers:

.. prompt:: bash device:~$, auto

    device:~$ docker ps

**Example Output**:

.. prompt:: text

     CONTAINER ID   IMAGE                               COMMAND                  CREATED          STATUS          PORTS                    NAMES
     a2d425490201   hub.foundries.io/<factory>/shellhttpd   "/usr/local/bin/http…"   20 minutes ago   Up 20 minutes   0.0.0.0:8080->8080/tcp   shellhttpd_httpd_1

Monitor the docker logs:

.. prompt:: bash device:~$, auto

    device:~$ docker logs --follow shellhttpd_httpd_1

**Example Output**:

.. prompt:: text

>      = Thu Mar 25 19:01:07 UTC 2021 =============================

Reviewer: This date and the one below on the same page are a month apart.  I don't think we want the user to assume this takes a month :)

Author: Shoud I change to:
`     = Thu MMM DD HH:MM:SS UTC YYYY =============================
`

Author: Or something like this:

`     = XXX XXX XX XX:XX:XX XXX XXXX =============================
`

Reviewer: Maybe let's drop these date lines.  They aren't important

Sharing Folder
^^^^^^^^^^^^^^

When you share a folder the base operating system and the 
running containers will be able to see and consume 

> the files inside it. This is one option to have then both to interact.

Reviewer 2: "to have **them** both" or "This is one option for the host machine and the device to interact"

> That is one of the ways to interact between the base operating system and the 
> running containers. When you share a folder, both will be able to see and consume 
> the files inside it.

Reviewer 3: ```suggestion
When you share a folder the base operating system and the 
running containers will be able to see and consume 
the files inside it. This is one option to have then both to interact.
```

> In the previous chapter, you have created and copied the configuration file to 

Reviewer 3: Maybe a cross reference would be better here (instead of "chapter")

> In the seccion ":ref:`tutorial-configuring-and-sharing-volumes-using-docker`", you have created and copied the configuration file to 

Reviewer 2: **section**

``/home/shellhttpd/`` folder. Purposefully leave the changes you just did because 
you will share the same folder to see what happens.

Edit the file ``docker-compose.yml`` adding the parameter volume:

.. prompt:: bash host:~$, auto

    host:~$ gedit shellhttpd/docker-compose.yml

**shellhttpd/docker-compose.yml**:

.. prompt:: text

     version: '3.2'
     
     services:
       httpd:
         image: hub.foundries.io/<factory>/shellhttpd:latest
     #    image: shellhttpd:1.0
         restart: always
         volumes:
           - /var/rootdirs/home/fio/shellhttpd:/home/shellhttpd/
         ports:
           - 8080:${PORT-8080}
         environment:
           MSG: "${MSG-Hello world}"

Above, you add the volumes parameter with the value  
``/var/rootdirs/home/fio/shellhttpd:/home/shellhttpd/``.
This means that the folder ``/var/rootdirs/home/fio/shellhttpd`` from your device 
will be mounted over the container folder ``/home/shellhttpd`` and the container 
folder content will be **overwritten** by the device folder content. 
Thus, as the device folder is empty, the ``shellhttpd.conf`` you copied in the 

> ``Dockerfile`` will **disappear** in this new setup.

Reviewer 2: Maybe rephrase this section changing the word 'device' as this can be misleading.
This means that the folder /var/rootdirs/home/fio/shellhttpd from your **host machine** will be mounted over the container folder /home/shellhttpd and the container folder content will be overwritten by the **host machine** folder content. Thus, as the **host machine** folder is empty, the shellhttpd.conf you copied in the Dockerfile will disappear in this new setup.

No need to keep this as bold, it was just to illustrate the words.

Reviewer: Agree this needs to be rephrased for clarity.

Check your changes, add, commit and push to the server:

.. prompt:: bash host:~$, auto

    host:~$ git status
    host:~$ git add shellhttpd/docker-compose.yml
    host:~$ git commit -m "Adding shared folder"
    host:~$ git push

> Make sure you received your update checking the latest target and the tab device 

Reviewer 3: ```suggestion
Make sure you received your update checking the latest target and the tab :guilabel:`Devices` 
```

> Make sure you received your update checking the latest target and the tab :guilabel:`Devices` 

Reviewer: - Add styling
- "update checking" -> "update by checking"
- "and the tab Devices" -> "on the Devices tab"
```
Make sure you received your update by checking the latest **Target** on the :guilabel:`Devices` tab 
```

in your Factory.

> Open one terminal connected to your device, check the running containers:

Reviewer: - Run-on sentence (add "and" between them)
```
Open one terminal connected to your device and check the running containers:
```

.. prompt:: bash device:~$, auto

>     device:~$ docker ps -a

Reviewer 4: If you want to confirm that it is running then don't use the `-a` because that checks all containers. It may confuse the user

**Example Output**:

.. prompt:: text

     CONTAINER ID   IMAGE                               COMMAND                  CREATED          STATUS          PORTS                    NAMES
     a2d425490201   hub.foundries.io/<factory>/shellhttpd   "/usr/local/bin/http…"   20 minutes ago   Up 20 minutes   0.0.0.0:8080->8080/tcp   shellhttpd_httpd_1

Monitor the docker logs:

.. prompt:: bash device:~$, auto

    device:~$ docker logs --follow shellhttpd_httpd_1

**Example Output**:

.. prompt:: text

     = Thu Mar 25 19:00:57 UTC 2021 =============================
     PORT=8080

>      MSG=Hello world

Reviewer: Not sure it would show "MSG=Hello world" at this point in the tutorial.

Author: It will because of:
```
    environment:
      MSG: "${MSG-Hello world}"
```

And the shared file wasn't created yet.

     punt!
     Terminated
     = Thu Mar 25 19:01:02 UTC 2021 =============================
     PORT=8080
     MSG=Hello world
     punt!
     Terminated
     = Thu Mar 25 19:01:07 UTC 2021 =============================
     PORT=8080
     MSG=Hello world

> Open a second terminal connected to your device, check for the folder 

Reviewer: - Run-on sentence
```
Open a second terminal connected to your device and check for the folder 
```

``/var/rootdirs/home/fio/shellhttpd``:

.. prompt:: bash device:~$, auto

    device:~$ ls /var/rootdirs/home/fio/shellhttpd/

The folder is empty and was automatically created when the Docker image was launched. 
Let’s create a new configuration file inside that folder and follow the logs 
from the first terminal:

.. prompt:: bash device:~$, auto

>     device:~$ sudo bash -c 'printf "MSG=\"Hello from shared folder\"" > /var/rootdirs/home/fio/shellhttpd/shellhttpd.conf'

Reviewer: Simplify(?) with `echo` -- maybe just consistency with my previous change
```
device:~$ sudo bash -c 'echo "MSG=\"Hello from shared folder\"" > /var/rootdirs/home/fio/shellhttpd/shellhttpd.conf'
```

And in the first terminal you should see the new ``MSG`` value:

**Example Output**:

.. prompt:: text

     = Tue Apr 27 01:01:02 UTC 2021 =============================
     PORT=8080

>      MSG=Hello world

Reviewer: Not sure this would show "MSG=Hello world" at this point in the tutorial.

     punt!
     Terminated
     = Tue Apr 27 01:01:07 UTC 2021 =============================
     PORT=8080
     MSG=Hello from shared folder

Just to confirm the change, test the container from an external device connected 
to the same network (e.g. your host machine: the same computer you use to access your device with ssh).

.. prompt:: bash host:~$, auto

    host:~$ curl <device IP>:8080

**Example Output**:

.. prompt:: text

     Hello from shared folder

At this point, you learned how to share a folder with the Docker container and 

> how to dynamically update the configuration manually.

Reviewer: - Can drop either "dynamically" or "manually" but don't need both:
```
how to manually update the configuration while the container is running.
```
